File: lc_soln/0115_min_stack/solution.cpp
class MinStack {
private:
    std::vector<int> data;
    std::vector<int> mins;
public:
    MinStack() {}

    void push(int val) {
        data.push_back(val);
        if (mins.empty()) {
            mins.push_back(val);
        } else {
            mins.push_back(std::min(getMin(), val));
        }
    }

    void pop() {
        data.pop_back();
        mins.pop_back();
    }

    int top() {
        return data.back();
    }

    int getMin() {
        return mins.back();
    }
};

int main() {
    return 0;
}

File: lc_soln/0115_min_stack/solution.rs
struct MinStack {
    data: Vec<i32>,
    mins: Vec<i32>,
}

impl MinStack {
    pub fn new() -> Self {
        Self {
            data: vec![],
            mins: vec![],
        }
    }

    pub fn push(&mut self, val: i32) {
        self.data.push(val);
        let min = if let Some(&curr) = self.mins.last() {
            val.min(curr)
        } else {
            val
        };
        self.mins.push(min);
    }

    pub fn pop(&mut self) {
        self.data.pop();
        self.mins.pop();
    }

    pub fn top(&self) -> i32 {
        self.data.last().copied().unwrap()
    }

    pub fn get_min(&self) -> i32 {
        self.mins.last().copied().unwrap()
    }
}

fn main() {}
